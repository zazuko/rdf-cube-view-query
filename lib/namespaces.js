import $rdf from '@zazuko/env'

const cube = $rdf.namespace('https://cube.link/')
const view = $rdf.namespace('https://cube.link/view/')

const {
  rdf,
  schema,
  sh,
  xsd,
} = $rdf.ns

export { cube, rdf, schema, sh, view, xsd }
