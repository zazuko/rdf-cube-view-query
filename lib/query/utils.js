import rdf from '@zazuko/env'

export function contains(ptr, target) {
  if (!target) {
    return ptr.terms.length > 0
  }

  if (target.term) {
    target = target.term
  }

  if (target.ptr && target.ptr.term) {
    target = target.ptr.term
  }

  return ptr.terms.some(term => term.equals(target))
}

export function distinct(...ptrs) {
  const ptr = ptrs[0]
  const termSet = rdf.termSet()

  for (const ptr of ptrs) {
    for (const term of ptr.terms) {
      termSet.add(term)
    }
  }

  const terms = [...termSet]

  if (terms.length === 0) {
    return ptr
  }

  return rdf.clownface({
    dataset: ptr._context[0].dataset,
    term: terms,
    graph: ptr._context[0].graph,
  })
}

export function toPropertyPath(ptr) {
  const list = ptr.list()

  if (!list) {
    return ptr.term
  }

  return [...list].map(property => property.term)
}
