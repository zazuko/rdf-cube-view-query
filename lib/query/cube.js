import rdf from '@zazuko/env'
import * as sparql from 'rdf-sparql-builder'
import * as ns from '../namespaces.js'

export function cubeQuery({ cube, graph = rdf.defaultGraph() } = {}) {
  const subject = rdf.variable('subject')

  const cubePattern = [sparql.bind(subject, `<${cube.value}>`)]
  const versionHistoryPattern = [[subject, ns.schema.hasPart, cube]]

  const patterns = [sparql.union([
    cubePattern,
    versionHistoryPattern,
  ])]

  return sparql.describe([subject])
    .from(graph)
    .where(patterns)
    .toString()
}
