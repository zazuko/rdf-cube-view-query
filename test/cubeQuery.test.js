import chai, { expect } from 'chai'
import { jestSnapshotPlugin } from 'mocha-chai-jest-snapshot'
import { cubeQuery } from '../lib/query/cube.js'
import * as ns from './support/namespaces.js'
import { cleanQuery } from './support/utils.js'

describe('query/cube', () => {
  chai.use(jestSnapshotPlugin())

  it('should create a DESCRIBE query for the cube and the version history', () => {
    const cube = ns.ex.cube
    const query = cubeQuery({ cube })

    expect(cleanQuery(query)).toMatchSnapshot()
  })

  it('should use the given named graph', () => {
    const cube = ns.ex.cube
    const graph = ns.ex.graph
    const query = cubeQuery({ cube, graph })

    expect(cleanQuery(query)).toMatchSnapshot()
  })
})
