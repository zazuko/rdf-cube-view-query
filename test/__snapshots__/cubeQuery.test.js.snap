// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`query/cube should create a DESCRIBE query for the cube and the version history 1`] = `
"DESCRIBE ?subject WHERE {
  { BIND(<http://example.org/cube> AS ?subject) }
  UNION
  { ?subject <http://schema.org/hasPart> <http://example.org/cube>. }
}"
`;

exports[`query/cube should use the given named graph 1`] = `
"DESCRIBE ?subject FROM <http://example.org/graph>
WHERE {
  { BIND(<http://example.org/cube> AS ?subject) }
  UNION
  { ?subject <http://schema.org/hasPart> <http://example.org/cube>. }
}"
`;
